### ISSUE TITLE
Set up Webhooks in the config file

### ISSUE THREAD

**Issue by NatoNathan** [supabase/cli#1190]
At the moment you have to manually setup database webhooks. You also need to redo the setup when you reset the DB. This becomes a problem when collaborating with others and sharing projects. It make it harder to provide a single command to setup and configure a project locally or remote.

It would be nice to be able to use the `config.toml` file to define webhooks that are automatically  setup when you start or reset a project. Maybe something like this?

```toml
[[db.webhooks]]
name = "my_webhook"
table= "auth.users"
operation = ["INSERT", "UPDATE", "DELETE"]
url = "http://localhost:3000/webhook"
params = { key = "value" }
headers = { key = "value" }
```

You can use SQL to make custom webhooks with triggers. but then you can't easily use environment variables for things like the url, params or headers.

**Comment by sweatybridge** [comment]
An alternative is to create web hooks from studio UI and save them as as a new migration. 

```bash
supabase db diff --schema supabase_functions -f my_webhook
```

Would the above work for you?

**Comment by NatoNathan** [comment]
The important thing is to be able to set a webhook like this

```toml
[[db.webhooks]]
name = "my_webhook"
table= "auth.users"
operation = ["INSERT", "UPDATE", "DELETE"]
url = "env(API_URL)"
params = { Authorization = "env(JWT)" }
headers = { key = "value" }
```

where `API_URL` and `JWT` change between local, staging and prod. otherwise I would be happy to use SQL to define the trigger.

**Comment by duyleekun** [comment]
Would `localhost` works in your case @NatoNathan

**Comment by NatoNathan** [comment]
> Would `localhost` works in your case @NatoNathan

Not really, I need to be able to target the docker host when running supabase locally and need to target supabase functions when we deploy it.

**Comment by bombillazo** [comment]
Yup, supabase hardcodes the webhook URL into the trigger definition, so any migration breaks the environment this was not build for...

**Comment by marcfrankel** [comment]
Jumping in to say this is also something we could use. It's hard to use db webhooks right now because the URL can't be dynamic based on environment variables

**Comment by rafaelmagalhaes** [comment]
has anyone come up with a temp solutions for this, facing some issues between development and production environments

**Comment by jmwind** [comment]
@sweatybridge I'm open to contributing a PR for to allow webhook config in the config.toml file. It's such a pain for local dev now that we depend on a few webhooks. 

I was looking thru the CLI source and wonder if you could give me a pointer. I was going to add a new section to the template as defined in `https://github.com/supabase/cli/blob/main/internal/utils/config.go` but then wasn't sure what you suggest in terms of where to read the config and create the webhook db records into the webhooks table. Is there an example where config from config.toml turns into a few upserts into the db?

I'm assuming that webhooks are stored in a table, but can't find it in the local schema. There's the hooks audit table in the `supabase_functions` table but can't find the associated definitions.

**Comment by bombillazo** [comment]
@jmwind i think webhooks are stored hardcoded into the triggers, not in tables.

**Comment by sweatybridge** [comment]
> url = "env(API_URL)"
params = { Authorization = "env(JWT)" }

I think it's tricky to add these configurable params via `config.toml` because we don't have any mechanism to push configs from local to remote project. What happens in case of conflict? Should it error, create new, or override existing webhooks?

Using data migrations would allow us to handle these cases using `on conflict` SQL clause so we don't need to reinvent the wheel. The blocker here is to have variable `api_url` and `anon_key` separate between local, staging, and prod.

We have a RFC internally to expose these as database configs, for eg. `select current_setting('supabase.project.url')` should return the corresponding `api_url` for local, staging, etc. I will try to get it prioritised soon.

**Comment by NatoNathan** [comment]
> The blocker here is to have variable `api_url` and `anon_key` separate between local, staging, and prod.
> 
> We have a [RFC](https://www.notion.so/supabase/Expose-project-vars-to-PG-instance-d9d7e2c7bf7a4e6983fddb877588c0c6) internally to expose these as database configs, for eg. `select current_setting('supabase.project.url')` should return the corresponding `api_url` for local, staging, etc. I will try to get it prioritised soon.

That definitely sounds like a really nice approach.

Another approach that could potential be implemented before that, is have the CLI generate and run the same queries that the studio runs to make and update webhooks. Although this would only fix this scenario. I have been meaning to investigate this solution and maybe even open a PR for a while but haven’t had the time.

**Comment by anthonyGustave** [comment]
> > url = "env(API_URL)"
> > params = { Authorization = "env(JWT)" }
> 
> ~I think it's tricky to add these configurable params via `config.toml` because we don't have any mechanism to push configs from local to remote project. What happens in case of conflict? Should it error, create new, or override existing webhooks?~
> 
> ~Using data migrations would allow us to handle these cases using `on conflict` SQL clause so we don't need to reinvent the wheel.~ nvm I was confused.
> 
> > webhooks are stored hardcoded into the triggers, not in tables.
> 
> This is correct.
> 
> The blocker here is to have variable `api_url` and `anon_key` separate between local, staging, and prod.
> 
> We have a [RFC](https://www.notion.so/supabase/Expose-project-vars-to-PG-instance-d9d7e2c7bf7a4e6983fddb877588c0c6) internally to expose these as database configs, for eg. `select current_setting('supabase.project.url')` should return the corresponding `api_url` for local, staging, etc. I will try to get it prioritised soon.

Just to be sure to understand, today there is nothing that allow to get variables using "select current_setting('supabase.project.url')" ?

**Comment by anthonyGustave** [comment]
I'm trying to do this approach bc current_setting is not available rn but it's not working:

```
SET search_path TO supabase_functions, public;

CREATE OR REPLACE FUNCTION new_mission_webhook()
RETURNS TRIGGER LANGUAGE plpgsql
SECURITY DEFINER
AS $$ 
DECLARE
    target_url TEXT;
    service_role_key TEXT;
    headers JSONB;
    payload JSONB;
    response TEXT;
BEGIN
    -- Fetch URL and service role key from the configuration table
    SELECT config_value INTO target_url FROM config.app_config WHERE config_key = 'api_url';
    SELECT config_value INTO service_role_key FROM config.app_config WHERE config_key = 'service_role_key';

    payload := '{}'::jsonb;

    -- Concatenate the additional path to the URL
    target_url := target_url || '/functions/v1/syncAirtable/supabase-to-airtable';

    -- Create headers
    headers := '{"Content-Type":"application/json"}'::jsonb;
    headers := jsonb_set(headers, '{Authorization}', ('"Bearer ' || service_role_key || '"')::jsonb);
    
    RAISE LOG 'Target URL: %', target_url;
    RAISE LOG 'Service Role Key: %', service_role_key;
    -- Debugging
    RAISE LOG 'Headers: %', headers;
    RAISE LOG 'Payload: %', payload;


    PERFORM "supabase_functions"."http_request"(
        'http://localhost:3000',
        'POST',
        '{"Content-Type":"application/json"}',
        '{}',
        '1000'
    );
    RAISE LOG 'Response: %', response;

    RETURN NEW;
END;

$$;

DROP TRIGGER IF EXISTS mission_webhook ON "public"."missions";
CREATE TRIGGER "mission_webhook" AFTER INSERT OR UPDATE OR DELETE
ON "public"."missions" FOR EACH ROW
EXECUTE FUNCTION new_mission_webhook();

ALTER FUNCTION new_mission_webhook() OWNER TO postgres;
```
But i'm getting:
function supabase_functions.http_request(unknown, unknown, unknown, unknown, unknown) does not exist

:(

**Comment by jdgamble555** [comment]
The staging / dev issue could be solved with this:

https://github.com/supabase/cli/issues/1518

And just run Supabase based on your desired env file.

J

**Comment by anthonyGustave** [comment]
I coded something hoping that it could help you. What is sad is that i don't use  supabase webhook in it but you need to create a new schema config and secure the schema. After connect the trigger to this code:

`CREATE OR REPLACE FUNCTION public.new_sync_webhook()
 RETURNS trigger
 LANGUAGE plpgsql
 SECURITY DEFINER
AS $function$ 
DECLARE
    target_url TEXT;
    service_role_key TEXT;
    headers JSONB;
    payload JSONB;
    response TEXT;
    operation_type TEXT;
    record_type TEXT;
BEGIN
    -- Fetch URL and service role key from the configuration table
    SELECT config_value INTO target_url FROM config.app_config WHERE config_key = 'api_url';
    SELECT config_value INTO service_role_key FROM config.app_config WHERE config_key = 'service_role_key';

    IF TG_OP = 'INSERT' THEN
        operation_type := 'INSERT';
    ELSIF TG_OP = 'UPDATE' THEN
        operation_type := 'UPDATE';
    ELSIF TG_OP = 'DELETE' THEN
        operation_type := 'DELETE';
    END IF;

    record_type := TG_TABLE_NAME;

    IF OLD IS NOT NULL THEN
        payload := jsonb_build_object(
            'table', record_type,
            'type', operation_type,
            'record', row_to_json(NEW),
            'old', row_to_json(OLD)
        );
    ELSE
        payload := jsonb_build_object(
            'table', record_type,
            'type', operation_type,
            'record', row_to_json(NEW)
        );
    END IF;

    -- Concatenate the additional path to the URL
    target_url := target_url || '/sync-airtable/to-airtable';

    SELECT extensions.http((
        'POST',
        target_url,
        ARRAY[extensions.http_header('Authorization','"Bearer ' || service_role_key || '"')],
        'application/json',
        payload::text
    )::extensions.http_request) INTO response;

    RETURN NEW;
EXCEPTION
    WHEN others THEN
        RAISE LOG 'HTTP request failed: %', SQLERRM;
END;

$function$
;`

**Comment by ykoitzsch** [comment]
any update on this?

**Comment by northfacejmb** [comment]
Also looking for updates on this.

**Comment by quememo** [comment]
This would be really useful

**Comment by lauri865** [comment]
Would love to be able to use environment variables in migration files (similar to config.toml) and define webhooks as migrations

**Comment by coryetzkorn** [comment]
Would love to be able to use environment variables in webhook HTTP request URLs. My requests currently get sent to a prod URL locally, leaving me with no way to test.

**Comment by tecoad** [comment]
This is a very much needed feature. Collaborating on this is awful though.

**Comment by jdgamble555** [comment]
The only thing I can think of is to run a custom script, for example `seed-vault.ts`, which adds `.env` variables to the vault locally. Then use them in your functions.

Related to this:

https://github.com/orgs/supabase/discussions/12813

But having this built into the migration file would be best.

J

**Comment by tecoad** [comment]
@jdgamble555 from what i undestand, I am able to use these secrets stored in the vault from the trigger or functions. Right?
But not able to find any documentation with that.
I guess i have to create function which returns the secret, then call this function from the trigger right?


Somethin like this
```
CREATE TRIGGER "Profile update"
  AFTER UPDATE ON "public"."profiles"
  FOR EACH ROW
  EXECUTE FUNCTION "supabase_functions"."http_request"(get_webhook_url(), '{"Content-Type":"application/json"}', '{"PARAMETER_NAME":"PARAMETER_VALUE"}', '1000');
```

But not able to make this work. Am i heading to the right path?

**Comment by jdgamble555** [comment]
I've never tested any of this so I couldn't tell you at this point, just theoretical.

**Comment by ric-cardo** [comment]
i was able to get it working using https://valentinprugnaud.dev/articles/setup-webhook-signature-with-supabase

**Comment by aoztanir** [comment]
Any update here?

**Comment by AntonGrafton** [comment]
Any updates?

**Comment by ollebergkvist** [comment]
Need this...

**Comment by clemwo** [comment]
> i was able to get it working using https://valentinprugnaud.dev/articles/setup-webhook-signature-with-supabase

link 404s unfortunately.

**Comment by rymawby** [comment]
> > i was able to get it working using https://valentinprugnaud.dev/articles/setup-webhook-signature-with-supabase
> 
> link 404s unfortunately.

I think this is the correct link mentioned above: https://blog.valentinprugnaud.dev/setup-webhook-signature-with-supabase

**Comment by coltonpomeroy** [comment]
My team could really use this feature.

**Comment by blytheaw** [comment]
We are also looking for this capability

**Comment by harrisrobin** [comment]
Just tried supabase branches and it seems like every new branch fails because it doesn't have database webhooks enabled by default. So if you need webhooks, it breaks branching or has anyone found a decent solution for now?

**Comment by sweatybridge** [comment]
@harrisrobin we are in the middle of documenting this: you can enable webhooks on branches by adding the following block to config.toml.

```
[experimental.webhooks]
enabled = true
```

**Comment by haeniya** [comment]
> We have a [RFC](https://www.notion.so/supabase/Expose-project-vars-to-PG-instance-d9d7e2c7bf7a4e6983fddb877588c0c6) internally to expose these as database configs, for eg. `select current_setting('supabase.project.url')` should return the corresponding `api_url` for local, staging, etc. I will try to get it prioritised soon.

@sweatybridge what's the status with this?

**Comment by tecoad** [comment]
@sweatybridge Any update on this?

How do we set webhook urls in per environment?

**Comment by icopp** [comment]
Also checking on this for the same reason. Currently there's no indication how to actually do any of this stuff without hardcoding everything for the production environment.

**Comment by sweatybridge** [comment]
Sorry for the lack of updates. As the postgres team has other priorities, I took a stab at this by supporting vault secrets in config.toml. You can try the following approach using cli beta.

1. Declare vault secrets in config.toml

```toml
[db.vault]
my_url = "http://localhost:3000"
my_token = "env(my_key)"
```

`my_key` should be defined in your `.env.local` file

2. Access secrets from any migration file

```sql
create table if not exists my_table (name text);

DO $$DECLARE
  url text;
  token text;
  headers jsonb;
BEGIN
  -- Fetch the webhook URL and token from vault
  select decrypted_secret into url from vault.decrypted_secrets where name = 'my_url';
  select decrypted_secret into token from vault.decrypted_secrets where name = 'my_token';
  -- Create webhook using secret url and token
  headers = jsonb_build_object (
    'Content-type', 'application/json',
    'Authorization', 'Bearer: ' || token
  );
  create or replace trigger "my_webhook" after insert
  on "public"."my_table" for each row
  execute function supabase_functions.http_request (
    url, 'POST', headers, '{}', '1000'
  );
END$$;
```

3. Run `supabase start` to verify webhooks locally.

You can then commit the migration in git, and set the `my_key` env var in your CI / CD environment to create the same webhook but using production credentials.

**Comment by bobzoller** [comment]
This is really cool and really close, but I had to wrap the `create or replace trigger` statement in an `execute` so that the variables `url` and `headers` were present at trigger-time, otherwise it threw a cryptic json error that meant the variable `headers` did not exist.

```
execute 'create or replace trigger "my_webhook" after insert
  on "public"."my_table" for each row
  execute function supabase_functions.http_request (
    ''' || url || ''', ''POST'', ''' || headers || ''', ''{}'', ''1000''
  );';
```

I'm pretty new here... does that sound right?

**Comment by sweatybridge** [comment]
@bobzoller yes that sounds alright. Another workaround is to wrap it with `DO ... BEGIN` block, similar to my example.

**Comment by bobzoller** [comment]
Hrm.  I had the `DO ... BEGIN` block just like your example, but the variable `headers` was not present at webhook-time. 🤷  perhaps I was _doing it wrong_!

**Comment by bobzoller** [comment]
Thank you so much for the follow-through here!  Seems my postgres experience ends at triggers and functions, so I appreciate the help.  This is way better (as you know) since the vault secrets will be read at webhook time instead of at migration time... letting the secrets evolve over time without needing a new migration. 💯

**Comment by geartvd** [comment]
> Sorry for the lack of updates. As the postgres team has other priorities, I took a stab at this by supporting vault secrets in config.toml. You can try the following approach using cli beta.
> 
> 1. Declare vault secrets in config.toml
> 
> [db.vault]
> my_url = "http://localhost:3000"
> my_token = "env(my_key)"
> `my_key` should be defined in your `.env.local` file
> 
> 2. Access secrets from any migration file
> 
> create table if not exists my_table (name text);
> 
> create or replace function my_webhook ()
>   returns trigger
>   security definer
>   language 'plpgsql'
> as $$
> declare
>   url text;
>   token text;
>   request_id bigint;
> begin
>   -- Get the webhook URL and token from vault
>   select decrypted_secret into url from vault.decrypted_secrets where name = 'webhook_url' limit 1;
>   select decrypted_secret into token from vault.decrypted_secrets where name = 'webhook_token' limit 1;
> 
>   -- Send the webhook request
>   select http_post into request_id from net.http_post(
>     url,
>     jsonb_build_object(
>       'old_record', OLD,
>       'record', NEW,
>       'type', TG_OP,
>       'table', TG_TABLE_NAME,
>       'schema', TG_TABLE_SCHEMA
>     ),
>     '{}',
>     jsonb_build_object(
>       'Content-Type', 'application/json',
>       'Authorization', 'Bearer ' || token
>     ),
>     '1000' -- timeout in ms
>   );
> 
>   -- Insert the request ID into the Supabase hooks table
>   insert into supabase_functions.hooks
>     (hook_table_id, hook_name, request_id)
>   values
>     (tg_relid, tg_name, request_id);
> 
>   return new;
> end;
> $$;
> 
> create trigger "my_webhook" after insert
> on "public"."my_table" for each row
> execute function my_webhook ();
> 3. Run `npx supabase@beta start` to verify webhooks locally.
> 
> You can then commit the migration in git, and set the `my_key` env var in your CI / CD environment to create the same webhook but using production credentials.

Thanks for this solution, works really well! One question though: How do you prevent the my_url from being overwritten on the production environment each time you run db push?

For example, suppose you have this:
Production vault: my_url=production_url
Local vault: my_url=local_url

And then in your config.toml
[db.vault]
my_url="env(my_url)"

But when you then run db push locally, the production vault my_url is overwritten with local_url, and you have to manually update it via the studio to production_url (easy to forget/make mistakes). Any way to prevent this?

**Comment by dewinterjack** [comment]
> Sorry for the lack of updates. As the postgres team has other priorities, I took a stab at this by supporting vault secrets in config.toml. You can try the following approach using cli beta.
> 
> 1. Declare vault secrets in config.toml
> 
> [db.vault]
> my_url = "http://localhost:3000"
> my_token = "env(my_key)"
> `my_key` should be defined in your `.env.local` file
> 
> 2. Access secrets from any migration file
> 
> create table if not exists my_table (name text);
> 
> create or replace function my_webhook ()
>   returns trigger
>   security definer
>   language 'plpgsql'
> as $$
> declare
>   url text;
>   token text;
>   request_id bigint;
> begin
>   -- Get the webhook URL and token from vault
>   select decrypted_secret into url from vault.decrypted_secrets where name = 'webhook_url' limit 1;
>   select decrypted_secret into token from vault.decrypted_secrets where name = 'webhook_token' limit 1;
> 
>   -- Send the webhook request
>   select http_post into request_id from net.http_post(
>     url,
>     jsonb_build_object(
>       'old_record', OLD,
>       'record', NEW,
>       'type', TG_OP,
>       'table', TG_TABLE_NAME,
>       'schema', TG_TABLE_SCHEMA
>     ),
>     '{}',
>     jsonb_build_object(
>       'Content-Type', 'application/json',
>       'Authorization', 'Bearer ' || token
>     ),
>     '1000' -- timeout in ms
>   );
> 
>   -- Insert the request ID into the Supabase hooks table
>   insert into supabase_functions.hooks
>     (hook_table_id, hook_name, request_id)
>   values
>     (tg_relid, tg_name, request_id);
> 
>   return new;
> end;
> $$;
> 
> create trigger "my_webhook" after insert
> on "public"."my_table" for each row
> execute function my_webhook ();
> 3. Run `npx supabase@beta start` to verify webhooks locally.
> 
> You can then commit the migration in git, and set the `my_key` env var in your CI / CD environment to create the same webhook but using production credentials.

@sweatybridge What is the point in inserting to `supabase_functions.hooks`? I tried this and it does not add the webhook to the studio UI unless you do `FOR EACH ROW EXECUTE FUNCTION supabase_functions.http_request`.

**Comment by barryhoughton** [comment]
> Sorry for the lack of updates. As the postgres team has other priorities, I took a stab at this by supporting vault secrets in config.toml. You can try the following approach using cli beta.
> 
> 1. Declare vault secrets in config.toml
> 
> [db.vault]
> my_url = "http://localhost:3000"
> my_token = "env(my_key)"
> `my_key` should be defined in your `.env.local` file
> 
> 2. Access secrets from any migration file
> 
> create table if not exists my_table (name text);
> 
> create or replace function my_webhook ()
>   returns trigger
>   security definer
>   language 'plpgsql'
> as $$
> declare
>   url text;
>   token text;
>   request_id bigint;
> begin
>   -- Get the webhook URL and token from vault
>   select decrypted_secret into url from vault.decrypted_secrets where name = 'webhook_url' limit 1;
>   select decrypted_secret into token from vault.decrypted_secrets where name = 'webhook_token' limit 1;
> 
>   -- Send the webhook request
>   select http_post into request_id from net.http_post(
>     url,
>     jsonb_build_object(
>       'old_record', OLD,
>       'record', NEW,
>       'type', TG_OP,
>       'table', TG_TABLE_NAME,
>       'schema', TG_TABLE_SCHEMA
>     ),
>     '{}',
>     jsonb_build_object(
>       'Content-Type', 'application/json',
>       'Authorization', 'Bearer ' || token
>     ),
>     '1000' -- timeout in ms
>   );
> 
>   -- Insert the request ID into the Supabase hooks table
>   insert into supabase_functions.hooks
>     (hook_table_id, hook_name, request_id)
>   values
>     (tg_relid, tg_name, request_id);
> 
>   return new;
> end;
> $$;
> 
> create trigger "my_webhook" after insert
> on "public"."my_table" for each row
> execute function my_webhook ();
> 3. Run `npx supabase@beta start` to verify webhooks locally.
> 
> You can then commit the migration in git, and set the `my_key` env var in your CI / CD environment to create the same webhook but using production credentials.

I couldn't get this to work... nothing appears in my local Vault ;(

<img width="278" height="75" alt="Image" src="https://github.com/user-attachments/assets/87b9aa31-8dd7-453c-802a-0afe37a6c050" />

<img width="1424" height="361" alt="Image" src="https://github.com/user-attachments/assets/7b725665-4327-4b42-b267-dca1def7e789" />
